File: Trabajo Final/Adjuntos/road_construction.cpp
#include <bits/stdc++.h>
using namespace std;

// Componentes de la red

int V;
vector<vector<int>> capacities;
vector<vector<int>> adj;

void addEdge(int u, int v, int c) {
  //printf("U:%d V:%d C:%d\n",u,v,c);
  adj[u].push_back(v);
  adj[v].push_back(u);
  capacities[u][v] += c;
}

int findPath(int st, int nd) {
  queue<pair<int, int>> bfs;
  vector<int> parent(V, -1);
  vector<bool> visited(V, false);
  
  bfs.push(make_pair(st, INT_MAX));
  parent[st] = -1;
  visited[st] = true;
  while (!bfs.empty()) {
    pair<int, int> now = bfs.front();
    bfs.pop();
    if (now.first == nd) {
      int ix = now.first;
      while (ix != st) {
        capacities[parent[ix]][ix] -= now.second;
        capacities[ix][parent[ix]] += now.second;
        ix = parent[ix];
      }
      return now.second;
    }
    for (int v : adj[now.first]) {
      if (capacities[now.first][v] > 0) {
        int f = min(capacities[now.first][v], now.second);
        if (!visited[v]) {
          parent[v] = now.first;
          visited[v] = true;
          bfs.push(make_pair(v, f));
        }
      }
    }
  }
  return 0;
}

int maxFlow(int st, int nd) {
  int flow = 0;
  int pathFlow = findPath(st, nd);
  while (pathFlow != 0) {
    flow += pathFlow;
    pathFlow = findPath(st, nd);
  }
  return flow;
}

void findCycle(vector<bool> &inCycle, int N, vector<vector<int>> &adj) {
  stack<pair<int, int>> dfs;
  dfs.push(make_pair(0, INT_MAX));
  vector<int> parent(N, -1);

  while (!dfs.empty()) {
    int u = dfs.top().first;
    int pt = dfs.top().second;
    dfs.pop();
    
    if (parent[u] == -1) {
      parent[u] = pt;
    } else {
      parent[u] = pt;
      int v = u;
      inCycle[v] = true;
      v = parent[v];
      while (u != v) {
        inCycle[v] = true;
        v = parent[v];
      }
      break;
    }

    for (int v : adj[u]) {
      if (v != pt) {
        dfs.push(make_pair(v, u));
      }
    }
  }
}

pair<int,int> getWorkerRoad(int worker, int N, map<int,int> materialToIndex, vector<int> workerList, vector<int> roads) {
  int matIndex = materialToIndex[workerList[worker]];
  for (int i = 0; i < N; ++i) {
    if (capacities[N+2+matIndex][i+2] == 1) {
      capacities[N+2+matIndex][i+2] = 0;
      return make_pair(i+1, roads[i]+1);
    }
  }
  return make_pair(0,0);
}

int main() {

  // Procesando la entrada
  int N, K;
  scanf("%d %d", &N, &K);
  
  // roads: ciudad i construye camino a roadList[i]
  vector<int> roads(N);
  // materialsForRoad: el camino de la ciudad i se puede construir con los materiales de materialsForRoad[i]
  vector<vector<int>> materialsForRoad(N);
  // adjCities: lista de adyacencias en grafo de ciudades y caminos
  vector<vector<int>> adjCities(N);

  for (int i = 0; i < N; ++i) {
    int M;
    scanf("%d %d", &roads[i], &M);
    --roads[i];
    adjCities[i].push_back(roads[i]);
    adjCities[roads[i]].push_back(i);
    materialsForRoad[i].resize(M);
    for (int j = 0; j < M; ++j) {
      scanf("%d", &materialsForRoad[i][j]);
    }
  }

  vector<int> workerList(K);
  // workers: cantidad de trabajadores que manipulan el material i
  vector<int> workers;
  // materialToIndex: Dado un material te da el indice para encontrarlo en workers
  map<int, int> materialToIndex;

  for (int i = 0; i < K; ++i) {
    scanf("%d", &workerList[i]);
    if (materialToIndex.count(workerList[i]) == 0) {
      materialToIndex[workerList[i]] = materialToIndex.size();
      workers.push_back(0);
    }
    ++workers[materialToIndex[workerList[i]]];
  }

  // Encontrar ciclo: cycle[i] es verdadero si la ciudad i pertenece al ciclo

  vector<bool> cycle(N, false);
  findCycle(cycle,N,adjCities);

  // Creacion de la red
  
  // Un vertice por cada ciudad, uno por cada material, un origen (0) y un destino (V-1), y uno extra para limitar el flujo para los nodos del ciclo (1)
  V = (N + 1) + materialToIndex.size() + 2;
  capacities.resize(V, vector<int>(V));
  adj.resize(V);

  int source = 0;
  int extraV = 1;
  int sink = V-1;

  // Conecto el vertice para limitar el flujo, y conecto las ciudades dentro del ciclo a este, mientras que las que esten fuera del ciclo a la fuente
  addEdge(source, extraV, count(cycle.begin(), cycle.end(), true) - 1);
  for (int i = 0; i < N; ++i) {
    if (cycle[i] && cycle[roads[i]]) {
      addEdge(extraV, i+2, 1);
    } else {
      addEdge(source, i+2, 1);
    }
    for (int mat : materialsForRoad[i]) {
      if (materialToIndex.count(mat)) {
        // Conecta el nodo de la ciudad i (i+2) con el de los posibles materiales
        addEdge(i+2, N+2+materialToIndex[mat], 1);
      }
    }
  }

  // Conecta los nodos de los materiales con el destino 
  for (int i = 0; i < materialToIndex.size(); ++i) {
    addEdge(N+2+i, sink, workers[i]);
  }
  
  // Calculo del flujo

  int mf = maxFlow(source, sink);
  
  // Obtencion de camino para cada trabajador

  if (mf < N - 1) {
    printf("-1");
  }
  else {
    for (int i = 0; i < K; ++i) {
      pair<int,int> road = getWorkerRoad(i,N,materialToIndex,workerList,roads);
      printf("%d %d\n",road.first, road.second);
    }
  }

  return 0;
}

File: Trabajo Final/Adjuntos/road_construction.py
import queue
import sys

def findPath(s, t):
    q = queue.Queue()
    prev = [-1 for i in range(V)]
    visited = [False for i in range(V)]

    q.put((s, None))
    visited[s] = True

    while(not q.empty()):
        node, cap = q.get()
        if (node == t):
            while (node != s):
                capacities[prev[node]][node] -= cap
                capacities[node][prev[node]] += cap
                node = prev[node]
            return cap
        for each in adj[node]:
            if capacities[node][each] > 0:
                if not visited[each]:
                    prev[each] = node
                    visited[each] = True
                    minCap = capacities[node][each] if (
                        cap == None) else min([cap, capacities[node][each]])
                    q.put((each, minCap))

    return 0

def maxFlow(s, t):
    flow = 0
    pathCapacity = findPath(s, t)
    while pathCapacity != 0:
        flow += pathCapacity
        pathCapacity = findPath(s, t)
    return flow

def addEdge(i, j,n):
    #print('U:',i,'V:',j,'N:',n)
    adj[i].append(j)
    adj[j].append(i)
    capacities[i][j] = n
    capacities[j][i] = 0


def getWorkerRoad(worker):
    matIndex = materialToIndex[workerList[worker]]
    for i in range(N):
        if capacities[N+2+matIndex][i+2] == 1:
            capacities[N+2+matIndex][i+2] = 0
            return(i+1, roads[i]+1)
    return(0, 0)

def findCycle(adj):
    inCycle = [False for i in range(N)]
    dfs = []
    dfs.append((0, None))
    prev = [-1 for i in range(N)]

    while (dfs):
        node, previous = dfs.pop()

        if (prev[node] == -1):
            prev[node] = previous
        else:
            prev[node] = previous
            v = node

            inCycle[v] = True
            v = prev[v]
            while (node != v):
                inCycle[v] = True
                v = prev[v]
            break

        for v in adj[node]:
            if (v != previous):
                dfs.append((v, node))

    return inCycle

# Procesando la entrada

N, K = map(int, sys.stdin.readline().split(' '))

# roads: ciudad i construye camino a roadList[i]
roads = []
# materialsForRoad: el camino de la ciudad i se puede construir con los materiales de materialsForRoad[i]
materialsForRoad = [[] for i in range(N)]
# adjCities: lista de adjacencias en grafo de ciudades y caminos
adjCities = [[] for i in range(N)]

for i in range(N):
    inp = list(map(int, sys.stdin.readline().split(' ')))

    destination = inp[0]-1
    
    roads.append(destination)
    adjCities[i].append(destination)
    adjCities[destination].append(i)

    materialsForRoad[i] = inp[2:]

workerList = list(map(int, sys.stdin.readline().split(' ')))
# workers: cantidad de trabajadores que manipulan el material i
workers = []
# materialToIndex: Dado un material te da el indice para encontrarlo en workers
materialToIndex = {}

for worker, material in enumerate(workerList):
    if material not in materialToIndex:
        materialToIndex[material] = len(materialToIndex)
        workers.append(0)

    workers[materialToIndex[material]] += 1

# Encontrar ciclo: cycle[i] es verdadero si la ciudad i pertenece al ciclo
cycle = findCycle(adjCities)

# Creacion de la red

# Un vertice por cada ciudad, uno por cada material, un origen (0) y un destino (V-1), y uno extra para limitar el flujo para los nodos del ciclo (1)
V = N + len(materialToIndex) + 3
capacities = [[-1 for i in range(V)] for j in range(V)]
adj = [[] for i in range(V)]

# Conecto el vertice para limitar el flujo, y conecto las ciudades dentro del ciclo a este, mientras que las que esten fuera del ciclo a la fuente
addEdge(0, 1,cycle.count(True)-1)
for i in range(N):
    if cycle[i] and cycle[roads[i]]:
        addEdge(1, i+2,1)
    else:
        addEdge(0, i+2,1)
    for material in materialsForRoad[i]:
        if material in materialToIndex:
            addEdge(i+2, N+2+materialToIndex[material], 1)

for i in range(len(materialToIndex)):
    addEdge(N+2+i, V-1, workers[i])

# Calculo del flujo
flow = maxFlow(0, V-1)
if (flow < N - 1):
    print(-1)
else:
    for worker in range(K):
        a, b = getWorkerRoad(worker)
        sys.stdout.write(str(a)+' '+str(b)+'\n')
